### PR TITLE
Shut down spark callback server when spark autologging is disabled or spark session is shut down

### LINKED ISSUE #8502: [BUG] with spark autolog, the Python script never finishes and hangs
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.3.2
- Tracking server: 2.3.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 13.3.1
- **Python version**: 3.11.3

### Describe the problem

As soon we call mlflow.spark.autolog() the Python script never finishes and hangs.
If I call spark.sparkContext._gateway.close() it closes properly the spark context and the script finishes.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
from pyspark.sql import SparkSession

spark = SparkSession.builder.config("spark.jars.packages", "org.mlflow:mlflow-spark:2.3.2").getOrCreate()
import mlflow
mlflow.spark.autolog() # work if we comment autolog

#spark.sparkContext._gateway.close() # work if we close the gateway

print("DONE")
```

### Stack trace

```
REPLACE_ME
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Related Issues/PRs

 #xxx
Closes https://github.com/mlflow/mlflow/issues/8502

## What changes are proposed in this pull request?

Shutdown spark callback server when spark autologging is disabled or spark session is shut down.

If we do not shut down spark callback server, even when spark session is shut down, the callback server is still running in background and it blocks python process exiting when python main thread exits.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by BenWilson2 — APPROVED** [review]
LGTM!

File: mlflow/_spark_autologging.py [hunk 1]
```diff
@@ -97,6 +97,16 @@ def _set_run_tag(run_id, path, version, data_format):
     client.set_tag(run_id, _SPARK_TABLE_INFO_TAG_NAME, new_tag_value)
 
 
+def _stop_listen_for_spark_activity(spark_context):
+    gw = spark_context._gateway
+    try:
+        gw.shutdown_callback_server()
+    except Exception as e:
+        _logger.warning(
+            "Failed to shut down Spark callback server for autologging: %s", str(e)
```
Comment (harupy) [inline mlflow/_spark_autologging.py:106]:
```suggestion
            "Failed to shut down Spark callback server for autologging: %s", e
```

Comment (harupy) [reply mlflow/_spark_autologging.py:106]: Do we need `for autologging`?

File: mlflow/spark.py [hunk 2]
```diff
@@ -999,9 +1002,14 @@ def autolog(disable=False, silent=False):  # pylint: disable=unused-argument
...
     """
     from mlflow.utils._spark_utils import _get_active_spark_session
-    from mlflow._spark_autologging import _listen_for_spark_activity
+    from mlflow._spark_autologging import (
+        _listen_for_spark_activity,
+        _stop_listen_for_spark_activity,
+    )
     from pyspark.sql import SparkSession
-    from pyspark import SparkContext
+    import atexit
```
Comment (harupy) [inline mlflow/spark.py:1010]: Does this need to be installed here?

Comment (harupy) [reply mlflow/spark.py:1010]: https://github.com/mlflow/mlflow/pull/8503 disallows imports like this one

File: mlflow/spark.py [hunk 3]
```diff
@@ -1010,9 +1018,25 @@ def __init__(original, self, *args, **kwargs):
...
+        sc = active_session.sparkContext
 
-        _listen_for_spark_activity(sc)
+        if disable:
+            _stop_listen_for_spark_activity(sc)
+        else:
+            _listen_for_spark_activity(sc)
+
+    if not _atexit_hook_installed:
+        # Register an "atexit" hook to disable spark autologging,
```
Comment (harupy) [inline mlflow/spark.py:1038]:
```suggestion
        # Register a hook to disable spark autologging,
```

File: mlflow/spark.py [hunk 4]
```diff
@@ -1010,9 +1018,25 @@ def __init__(original, self, *args, **kwargs):
...
+        # so that before python process exits, we can shut down
+        # spark callback server properly.
+        atexit.register(mlflow.spark.autolog, disable=True)
```
Comment (harupy) [inline mlflow/spark.py:1041]: Can we explicitly shutdown spark callback server?

File: mlflow/spark.py (hunk 3 above)
Comment (harupy) [reply mlflow/spark.py:1038]: Never mind this comment.

Comment (harupy) [inline mlflow/spark.py:1038]: Do we need to disable spark autologging? We just want to shutdown spark, right?

File: mlflow/spark.py [hunk 5]
```diff
@@ -1010,9 +1018,30 @@ def __init__(original, self, *args, **kwargs):
...
+        else:
+            _listen_for_spark_activity(sc)
+
+    if not _atexit_hook_registered:
+        # Register an "atexit" hook to disable spark autologging,
+        # so that before python process exits, we can shut down
+        # spark callback server properly.
+        def shut_down_spark_callback_server():
+            active_session = _get_active_spark_session()
+            if active_session is not None:
```
Comment (harupy) [inline mlflow/spark.py:1043]:
```suggestion
            if active_session := _get_active_spark_session():
```

File: mlflow/spark.py [hunk 6]
```diff
@@ -1010,9 +1018,30 @@ def __init__(original, self, *args, **kwargs):
...
+        sc = active_session.sparkContext
 
-        _listen_for_spark_activity(sc)
+        if disable:
+            _stop_listen_for_spark_activity(sc)
+        else:
+            _listen_for_spark_activity(sc)
+
+    if not _atexit_hook_registered:
+        # Register an "atexit" hook to disable spark autologging,
```
Comment (harupy) [inline mlflow/spark.py:1038]: Is this comment still accurate?

File: mlflow/spark.py (hunk 5 above)
Comment (harupy) [reply mlflow/spark.py:1043]: You can define this function in the top-level.

**Comment by WeichenXu123** [comment]
I just tested atexit hook but I found it does not work,
because atexit hook is called only after all non-daemon background threads exit,
but the spark callback server will generate some non-daemon background threads which blocks triggering atexit hook when main thread exits,
so that I removed atexit hook code,
user can add `sparkSession.stop()` at the end of main thread code to make the python process exit properly.

File: mlflow/spark.py [hunk 7]
```diff
@@ -933,6 +934,9 @@ def predict(self, pandas_df):
         ]
 
 
+_atexit_hook_registered = False
```
Comment (harupy) [inline mlflow/spark.py:937]: Is this used?

File: mlflow/spark.py [hunk 8]
```diff
@@ -999,9 +1000,13 @@ def autolog(disable=False, silent=False):  # pylint: disable=unused-argument
...
     from mlflow.utils._spark_utils import _get_active_spark_session
-    from mlflow._spark_autologging import _listen_for_spark_activity
+    from mlflow._spark_autologging import (
+        _listen_for_spark_activity,
+        _stop_listen_for_spark_activity,
+    )
     from pyspark.sql import SparkSession
-    from pyspark import SparkContext
+
+    global _atexit_hook_registered
```
Comment (harupy) [inline mlflow/spark.py:1009]: Is this used?

**Pull request merged**